src/components/Model3D.tsx

```diff
@@ -1,29 +1,65 @@
+// TODO is this ideal?
+declare global {
+  interface Window {
+    viewer: Cognite3DViewer;
+    THREE: any;
+    model: any;
+  }
+}
 import React from 'react';
 import { connect } from 'react-redux';
-import PropTypes from 'prop-types';
-import { Model3DViewer } from '@cognite/gearbox';
-import { THREE } from '@cognite/3d-viewer';
+import { Model3DViewer, CacheObject, Callback } from '@cognite/gearbox';
+import { THREE, Cognite3DViewer, Cognite3DModel } from '@cognite/3d-viewer';
 import { message, Slider } from 'antd';
 import * as sdk from '@cognite/sdk';
 import LoadingScreen from './LoadingScreen';
-import { fetchNode } from '../modules/threed';
-
 import {
   fetchMappingsFromNodeId,
   selectAssetMappings,
-  AssetMappings,
   fetchMappingsFromAssetId,
+  AssetMappingState
 } from '../modules/assetmappings';
-import { selectFilteredSearch, Filters } from '../modules/filters';
+import { selectFilteredSearch } from '../modules/filters';
+import { Dispatch, bindActionCreators } from 'redux';
+import { RootState } from '../reducers/index';
+import { Asset } from '@cognite/sdk';
+import { ExtendedAsset } from '../modules/assets';
+import { ProgressObject } from './LoadingScreen';
+import { SliderValue } from 'antd/lib/slider';
+import { fetchNode } from '../modules/threed';
 
-class Model3D extends React.Component {
-  state = { hasWarnedAboutNode: {} };
+type Props = {
+  modelId: number;
+  revisionId: number;
+  cache?: CacheObject;
+  nodeId?: number;
+  onAssetIdChange: (assetId: number) => void;
+  doFetchMappingsFromNodeId: typeof fetchMappingsFromNodeId;
+  doFetchMappingsFromAssetId: typeof fetchMappingsFromAssetId;
+  doFetchNode: typeof fetchNode;
+  assetMappings: AssetMappingState;
+  filteredSearch: { items: ExtendedAsset[] };
+};
 
-  currentColoredNodes = [];
+type State = {
+  hasWarnedAboutNode: { [key: string]: boolean };
+  nodeId?: number;
+  progress?: ProgressObject;
+};
 
-  cache = {};
+class Model3D extends React.Component<Props, State> {
+  static defaultProps = {
+    assetMappings: { byNodeId: {}, byAssetId: {} },
+    nodeId: undefined,
+    cache: undefined
+  };
+  readonly state: Readonly<State> = { hasWarnedAboutNode: {} };
 
-  componentDidUpdate(prevProps, prevState) {
+  currentColoredNodes: number[] = [];
+  model?: Cognite3DModel = undefined;
```

Comment:
Any way we could rewrite this class to be guaranteed to have a model and viewer? I.e. by passing them to the constructor or similar.

Reply:
sadly because of React's lifecycle we cant expect anything else in the constructor and need to wait for the `onReady` callback